File: webui/features/train.py
"""Train: ``train.py -c <config> [-t <ckpt>]``, mirroring scripts/dist_train.sh."""

from pathlib import Path

from .base import (
    Feature,
    Field,
    JobSpec,
    config_path,
    existing_file,
    gpu_env,
    launcher,
    positive_int,
    python_executable,
    runtime_rows,
    timestamp,
)


class TrainFeature(Feature):
    name = "train"
    label = "Train"
    description = "Start a training run from scratch or from a tuning checkpoint."
    fields = [
        Field(
            "config",
            "Config (-c)",
            kind="choice",
            source="configs",
            value="configs/dome/Dome-M-AEA.yml",
            prefer="AEA",
        ),
        Field(
            "checkpoint",
            "Tuning checkpoint (-t, optional)",
            kind="choice",
            source="checkpoints",
            optional=True,
            empty_label="(none / from scratch)",
        ),
        *runtime_rows(port=7789),
    ]

    def build(self, params):
        config = config_path(params)
        tuning = existing_file(params, "checkpoint", "tuning checkpoint (-t)", required=False)
        seed = positive_int(params, "seed", 42)
        nproc = positive_int(params, "nproc", 1)
        port = positive_int(params, "port", 7789)

        outdir = f"output/{Path(config).stem}/{timestamp()}"
        cmd = launcher(python_executable(params), nproc, port)
        cmd += ["-c", config, "--seed", str(seed), "--output-dir", outdir]
        if tuning:
            cmd += ["-t", tuning]

        meta = {"feature": self.name, "config": config, "outdir": outdir, "cmd": " ".join(cmd)}
        return JobSpec(cmd, env=gpu_env(params), meta=meta)

File: webui/static/js/features/train.js
// Train tab -- pairs with webui/features/train.py

export default {
  name: "train",
  fields: [
    { name: "config", label: "Config (-c)", type: "select", source: "configs", prefer: "AEA" },
    {
      name: "checkpoint",
      label: "Tuning checkpoint (-t, optional)",
      type: "select",
      source: "checkpoints",
      empty: "(none / from scratch)",
    },
    [
      { name: "gpus", label: "GPUs (CUDA_VISIBLE_DEVICES)", value: "0", placeholder: "e.g. 0,1" },
      { name: "seed", label: "Seed", value: "42" },
    ],
    [
      { name: "nproc", label: "nproc_per_node", value: "1" },
      { name: "port", label: "master_port", value: "7789" },
    ],
  ],
};
